enterprise_gateway/services/kernels/remotemanager.py

```diff
@@ -189,6 +192,57 @@ async def start_kernel(self, *args, **kwargs):
         self.parent.kernel_session_manager.create_session(kernel_id, **kwargs)
         return kernel_id
 
+    async def restart_kernel(self, kernel_id):
+        kernel = self.get_kernel(kernel_id)
+        self.log.debug(f"Current value of the kernel.restarting: {kernel.restarting}")
+        if kernel.restarting:  # assuming duplicate request.
+            await self.wait_for_restart_finish(kernel_id, "restart")
+            self.log.info(
+                f"Done with waiting for restart to complete. Current value of kernel.restarting: {kernel.restarting}. "
+                f"Skipping kernel restart."
+            )
+            return
+        self.log.info("Going ahead to process kernel restart request.")
+        try:
+            kernel.restarting = True  # Moved in out of RemoteKernelManager
+            await super().restart_kernel(kernel_id)
+        finally:
+            self.log.debug("Resetting kernel.restarting flag to False.")
+            kernel.restarting = False
+
+    async def shutdown_kernel(self, kernel_id, now=False, restart=False):
+        kernel = self.get_kernel(kernel_id)
+        self.log.debug(f"Current value of the Kernel Restarting: {kernel.restarting}")
+        if kernel.restarting:
+            await self.wait_for_restart_finish(kernel_id, "shutdown")
+        try:
+            await super().shutdown_kernel(kernel_id, now, restart)
+        except KeyError as ke:  # this is hit for multiple shutdown request.
+            self.log.exception(f"Exception while shutting Kernel: {kernel_id}: {ke}")
+
+    async def wait_for_restart_finish(self, kernel_id, action="shutdown"):
+        kernel = self.get_kernel(kernel_id)
+        start_time = float(time.time())  # epoc time
+        timeout = kernel.kernel_launch_timeout
+        poll_time = 1  # TODO can be read from a new config if need be.
```

Comment:
Let's tie this to an env at a minimum.  How about `EG_RESTART_FINISH_POLL_INTERVAL`?  A default of 1.0 seems reasonable.

Reply:
agree. I will  make this change.
- does EG_RESTART_STATUS_POLL_INTERVAL sounds better?
- will this also need a doc update for env variables?

Reply:
That name sounds good - thanks.  Yeah, we should probably document this.  I think placing it into the [_Environment variables that assist in troubleshooting_](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/operators/config-env-debug.html) section is a good spot.

Reply:
sure. ll make this change as well.